substrate/client/authority-discovery/src/worker/addr_cache.rs

```diff
@@ -33,16 +50,52 @@ pub(super) struct AddrCache {
 	/// it's not expected that a single `AuthorityId` can have multiple `PeerId`s.
 	authority_id_to_addresses: HashMap<AuthorityId, HashSet<Multiaddr>>,
 	peer_id_to_authority_ids: HashMap<PeerId, HashSet<AuthorityId>>,
+
+	on_change: Option<OnAddrCacheChange>,
+}
+impl AddrCache {
+	/// Clones all but the `on_change` handler (if any).
+	fn clone_content(&self) -> Self {
```

Comment:
`Box<dyn Fn(AddrCache) -> Result<(), Box<dyn fmt::Debug>> + 'static>` is not `Clone`, so this is a manual impl of cloning the relevant content, used in:

```rust
	fn notify_change_if_needed(&self) {
		if let Some(on_change) = &self.on_change {
			match (on_change)(self.clone_content()) {
				Ok(()) => {},
				Err(err) => {
					log::error!(target: super::LOG_TARGET, "Error while notifying change: {:?}", err);
				},
			}
		}
	}
```